### PR TITLE
Update runtime repo to use Preview 7 SDK

### LINKED ISSUE #65317: Unpin C# compiler version for repo
We're currently manually pinning the C# compiler version:
https://github.com/dotnet/runtime/blob/ce13258b04df862e4a2cd8a36045fa7491414d10/eng/Versions.props#L55-L56
as the version currently specified from dotnet/arcade is too old:
```
4.1.0-2.21609.7 (81831342)
```
We need to ensure the compiler bits are flowing appropriately and unpin.

### PR DESCRIPTION
**Pull request by agocke**

### CODE REVIEW

**Review by ViktorHofer — APPROVED** [review]
LGTM. Of course, before merging we need to go through the usual process of communicating the change.

**Comment by agocke** [comment]
@MichalStrehovsky @LakshanF  Looks like some NativeAOT libraries tests are failing after an SDK update. Ideas?

**Comment by MichalStrehovsky** [comment]
> @MichalStrehovsky @LakshanF Looks like some NativeAOT libraries tests are failing after an SDK update. Ideas?

Looking... one thing that's missing is adding handling of the new TrimMode to NativeAOT targets. But the failure mode doesn't look like that's the problem.

**Comment by MichalStrehovsky** [comment]
> > @MichalStrehovsky @LakshanF Looks like some NativeAOT libraries tests are failing after an SDK update. Ideas?
> 
> Looking... one thing that's missing is adding handling of the new TrimMode to NativeAOT targets. But the failure mode doesn't look like that's the problem.

It's from the new aggresse trimming default in Preview 6. xUnit is calling `PropertyInfo.GetValue` on `CollectionBehaviorAttribute.DisableTestParallelization` but we trimmed the getter (the setter was left because the attribute is applied on something and we need the setter to materialize the `Attribute` instance at runtime - the getter is fair game).

This is actually a very lucky catch. Only a couple of the tests fail. The rest look like this:

```
Running assembly:System.Runtime.Tests, Version=7.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
Finished System.Runtime.Tests, Version=7.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51

Tests run: 0, Errors: 0, Failures: 0, Skipped: 0. Time: 0.0130819s
```

The fix will be along the lines of adding `<TrimMode>partial</TrimMode>` to `eng\testing\tests.singlefile.targets`.

File: src/tests/nativeaot/Directory.Build.props
Comment (MichalStrehovsky) [inline src/tests/nativeaot/Directory.Build.props:4]: We'll probably need to do the opposite here - nativeaot tests expect trimming and require the (now default) TrimMode=full. So this line is technically not needed. But the rest of the coreclr test tree will want to keep the partial TrimMode.

File: src/coreclr/tools/aot/crossgen2/crossgen2.props
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/crossgen2/crossgen2.props:17]: Do we need this line if this is now the default?

Comment (agocke) [reply src/coreclr/tools/aot/crossgen2/crossgen2.props:17]: No, we don't, I was just being explicit.

**Comment by agocke** [comment]
Does NativeAOT understand `TrimMode=partial` or does that need to be added?

**Comment by MichalStrehovsky** [comment]
> Does NativeAOT understand `TrimMode=partial` or does that need to be added?

I think https://github.com/dotnet/runtime/pull/71117 is what's needed. Could be folded into this PR to avoid adding the TODO.

**Comment by agocke** [comment]
It looks like the installer tests are broken by https://github.com/dotnet/sdk/issues/26462. The test execution happens through Arcade, so we can't pass DOTNET_ROOT as is.

I see two options for workaround:

1. Add a capability to Arcade to set environment variables for the test process
2. Update to an SDK daily build that contains the fix.

I don't see any way to use the Preview 6 SDK otherwise.

**Comment by stephentoub** [comment]
> I think this should be updated to P7 as soon as that's released and we should merge ASAP

Yes.  FYI, we're going to need to update our C# compiler very soon to pick up changes from @cston around scoped, but at least the dotnet-pgo project (and maybe others) will fail to build with that compiler until we move to a newer SDK that has consistency between exposing the ByRefsFeature and having relevant APIs annotated as scoped.
cc: @AaronRobinsonMSFT

**Comment by ViktorHofer** [comment]
@dotnet/runtime-infrastructure as @agocke is out, does someone else want to drive this change now that P7 is out?

**Comment by jkoritzinsky** [comment]
I can take over on this

**Comment by jkoritzinsky** [comment]
We're seeing JIT asserts here in Importation when using the live JIT.

cc: @dotnet/jit-contrib can someone take a look at this?

**Comment by JulieLeeMSFT** [comment]
@kunalspathak PTAL.

**Comment by jakobbotsch** [comment]
This sounds suspiciously much like https://github.com/dotnet/roslyn/issues/62864.

**Comment by kunalspathak** [comment]
Yep, looking at the errors we are hitting

`assert(!"Unhandled must expand intrinsic, throwing PlatformNotSupportedException");`

**Comment by jkoritzinsky** [comment]
So I guess we need to update the Roslyn version to at least a version with the fix?

**Comment by AaronRobinsonMSFT** [comment]
This is an expected issue with the pgo project. I'll push up a fix.

**Comment by AaronRobinsonMSFT** [comment]
@MichalStrehovsky or @vitek-karas Any thoughts on the NativeAOT failures?

```
/_/src/xunit.execution/Extensions/ReflectionAbstractionExtensions.cs(92): Trim analysis error IL2070: ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Reflection.TypeInfo.DeclaredMethods.get'. The parameter 'type' of method 'ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/__w/1/s/src/libraries/System.Runtime.Intrinsics/tests/System.Runtime.Intrinsics.Tests.csproj]
##[error]/_/src/xunit.execution/Extensions/ReflectionAbstractionExtensions.cs(92,0): error IL2070: (NETCORE_ENGINEERING_TELEMETRY=Build) ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Reflection.TypeInfo.DeclaredMethods.get'. The parameter 'type' of method 'ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
/_/src/xunit.execution/Extensions/ReflectionAbstractionExtensions.cs(92): Trim analysis error IL2070: ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Reflection.TypeInfo.DeclaredMethods.get'. The parameter 'type' of method 'ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/__w/1/s/src/libraries/System.Runtime.Intrinsics/tests/System.Runtime.Intrinsics.Tests.csproj]
##[error]/_/src/xunit.execution/Extensions/ReflectionAbstractionExtensions.cs(92,0): error IL2070: (NETCORE_ENGINEERING_TELEMETRY=Build) ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Reflection.TypeInfo.DeclaredMethods.get'. The parameter 'type' of method 'ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
/_/src/xunit.execution/Extensions/ReflectionAbstractionExtensions.cs(92): Trim analysis error IL2067: ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Reflection.RuntimeReflectionExtensions.GetRuntimeMethods(Type)'. The parameter 'type' of method 'ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/__w/1/s/src/libraries/System.Runtime.Intrinsics/tests/System.Runtime.Intrinsics.Tests.csproj]
##[error]/_/src/xunit.execution/Extensions/ReflectionAbstractionExtensions.cs(92,0): error IL2067: (NETCORE_ENGINEERING_TELEMETRY=Build) ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Reflection.RuntimeReflectionExtensions.GetRuntimeMethods(Type)'. The parameter 'type' of method 'ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
/_/src/xunit.execution/Extensions/ReflectionAbstractionExtensions.cs(92): Trim analysis error IL2067: ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Reflection.RuntimeReflectionExtensions.GetRuntimeMethods(Type)'. The parameter 'type' of method 'ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/__w/1/s/src/libraries/System.Runtime.Intrinsics/tests/System.Runtime.Intrinsics.Tests.csproj]
##[error]/_/src/xunit.execution/Extensions/ReflectionAbstractionExtensions.cs(92,0): error IL2067: (NETCORE_ENGINEERING_TELEMETRY=Build) ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Reflection.RuntimeReflectionExtensions.GetRuntimeMethods(Type)'. The parameter 'type' of method 'ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
```

**Review by jaredpar — APPROVED** [review]

**Comment by MichalStrehovsky** [comment]
> @MichalStrehovsky or @vitek-karas Any thoughts on the NativeAOT failures?

This is WarnAsError in action.

Apparently we deleted what it means to SuppressTrimAnalysisWarnings in https://github.com/dotnet/linker/pull/2930 and NativeAOT no longer suppresses these. I've inlined what that means into the targets to work around and filed a 7.0 issue to fix it because this is customer-visible (https://github.com/dotnet/runtime/issues/73926).

I'm puzzled why we're seeing it here because the break merged into linker repo 10 days ago and there's no way that made it into Preview 7 (in fact, troubleshooting this took longer than I'm willing to admit because the .targets in P7 SDK still have this suppression but we're picking it up from a NuGet instead).

**Comment by MichalStrehovsky** [comment]
The Linux_musl x64 Debug failure is https://github.com/dotnet/runtime/issues/73930 (a test touched 4 hours ago, so I'm going with the "unrelated to what we're doing here" theory).

We're good to merge once llvmfullaot is done with its thing.

File: src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs
Comment (teo-tsirpanis) [inline src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs:77]: Won't that cause a GC hole? Same with above.

Comment (MichalStrehovsky) [reply src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs:77]: Good catch. We either need to make this a `ref struct` with a big "DO NOT CLEAN UP THE REF PART" comment, or use `fixed` to grab the pointer instead.

Comment (jakobbotsch) [reply src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs:77]: Yes it will. I'll push a fix.

Comment (jakobbotsch) [reply src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs:77]: > We either need to make this a ref struct with a big "DO NOT CLEAN UP THE REF PART" comment, or use fixed to grab the pointer instead.

I think we can make these static functions that take the owning struct by ref.

Comment (teo-tsirpanis) [reply src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs:77]: Can't we make it

```suggestion
            return MemoryMarshal.CreateSpan(ref _entries, (totalSize - (int)entriesOffset) / sizeof(LbrEntry64));
```

and add `[UnscopedRef]` to the method?

Comment (jakobbotsch) [reply src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs:77]: Yes we can, that's an even smaller changeset, I'll do that. Thanks for the suggestion.

**Comment by radical** [comment]
@jakobbotsch build failed

**Comment by jakobbotsch** [comment]
> @jakobbotsch build failed

Hmm strange, it builds fine from within VS locally. I suppose we cannot use this attribute yet in the project. Let me go back to my other fix.

File: src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs
Comment (AaronRobinsonMSFT) [reply src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs]: This isn't a GC hole in this case. It is actually correct.

Comment (AaronRobinsonMSFT) [reply src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs]: We were trying to keep this narrow due to all the changes in flight. The problem here is the project is referencing reference assemblies from Preview 7 to build and the annotations here are simply wrong. Once the repo moves to RC1, this can be removed and updated appropriately.

Comment (AaronRobinsonMSFT) [reply src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs]: i shouldn't say "correct", rather it is "safe.

Comment (AaronRobinsonMSFT) [reply src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs]: I really wish this hadn't been touched.

Comment (jakobbotsch) [reply src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs]: @AaronRobinsonMSFT I own this code -- I would prefer to have the code be correct by construction. I agree that it was fine before, but it was leaving an implicit assumption on how I am going to work with this code in the future.

Comment (AaronRobinsonMSFT) [reply src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs]: I see. This was something I was shepherding through for `ref` field support so it was in a state I knew to track. I will tag you if anything more comes up.

Comment (jakobbotsch) [reply src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs]: Ah okay, sorry about that, I did not mean to step on your toes here. 

> I will tag you if anything more comes up.

Thanks, I appreciate it.

Comment (AaronRobinsonMSFT) [reply src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs]: Since I've also got the owner here, can we get this moved to build against the live build? Building against the repo's SDK is the root issue here.

Comment (jakobbotsch) [reply src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs]: I can look into it, I'm not totally sure on the constraints about that (we are publishing dotnet-pgo as a nuget package, would that be affected?). I probably need to involve @davidwrighton here.
I assume it can be done separately?

Comment (AaronRobinsonMSFT) [reply src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs]: > I assume it can be done separately?

Absolutely. My only concern is this project is the only one I found that seems to use the repo's SDK and could be heavily impacted by complex language changes that force updates to the public API surface. This is a longer-term concern and just something to keep in mind as the language evolves to have more of these low-level friendly features.

**Comment by radical** [comment]
The wasm - AOT/EAT failures are present on `main` too, and are - https://github.com/dotnet/runtime/issues/73961 .
The wasm debugger test failure is - https://github.com/dotnet/runtime/issues/73957

**Comment by jkoritzinsky** [comment]
Linux x64 Debug leg failure is https://github.com/dotnet/runtime/issues/73930

**Comment by jkoritzinsky** [comment]
All failures are accounted for. Merging this in.

**Comment by ViktorHofer** [comment]
Thanks for getting this in everyone. @jkoritzinsky as usual, can you please send out the notice to the team including instructions on how to obtain the new SDK (for global installs)?

**Comment by jkoritzinsky** [comment]
Will do!

File: src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs
Comment (MichalStrehovsky) [reply src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs]: > My only concern is this project is the only one I found that seems to use the repo's SDK and could be heavily impacted by complex language changes that force updates to the public API surface

Everything under src/coreclr/tools builds against the repo's SDK. Building against live bits has build sequencing issues where one wouldn't be able to build crossgen2 without building libs first, or run crossgen2 without building the host.

If we want to switch to live, I would want to know what the VS debug experience for live built things looks like. I would still prefer we don't do that because I kind of like I don't need to build coreclr VM, host, or other stuff for NativeAOT development.

Comment (AaronRobinsonMSFT) [reply src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs]: > Building against live bits has build sequencing issues

Ah. Thanks @MichalStrehovsky. This is an important consideration to note. Perhaps this wouldn't justify the resulting dev-loop pain.

**Pull request merged**
